torch/distributed/_local_tensor/__init__.py

```diff
@@ -0,0 +1,749 @@
+from ast import Call
+
+
+"""
+A LocalTensor is a tensor subclass which simulates a tensor that is
+distributed across SPMD ranks.  A LocalTensor might be size N, but in fact
+there are world_size shards/replicas of it stored internally.  When you do a
+plain PyTorch operation on it, we apply the operation to each shard; when you
+do a collective, we do the mathematically equivalent operation on the local
+shards.  A LocalTensor is associated with a list of ranks which specify
+which ranks it holds local tensors for.
+
+NB, this is NOT a DataParallel like abstraction where you can run operations
+on multiple different GPUs. It is intended purely for *debugging* purposes,
+the overhead is almost certainly too high to keep eight GPUs (even the C++
+autograd needs multithreading to keep up!)  (It might potentially be possible
+to trace through this with torch.compile and then compile it with CUDA graphs
+but this is currently a non-goal.)
+
+In order to handle MPMD, we provide a helper decorator that allows you to
+run a function with no side effects for each LocalTensor shard and combine
+results back into LocalTensor or LocalIntNode.
```

Comment:
This isn't full MPMD though, it's a sort of limited divergence but all the ranks get back into synchrony by the time you get to LocalTensor/LocalIntNode. MPMD here would evoke PP which is not right.